Add schema inference

dtpeacock opened the pull request:
This resolves #60 , the schema is inferred in a similar way to how the JSON schema is inferred.

The public api has not changed other than adding the inferSchema option, e.g.:
    
    csvFile(carsFile, inferSchema = true)

    OPTIONS (path "path", inferSchema "true")

I have pulled the tokenizing code into a common method as the schema has to be inferred on the tokenized data.

Currently this uses the entire dataset to discover the schema, I could also add on a sampling `samplingRatio` option as in the JSON schema inference?

mohitjaggi commented on src/main/scala/com/databricks/spark/csv/util/InferSchema.scala:
```diff
@@ -0,0 +1,73 @@
+
+import scala.util.control.Exception._
+
+private[csv] object InferSchema {
```
thanks. schema inference would be awesome...i wonder what the performance would be like, as exceptions can be expensive. on the other hand you can see in bigdf code https://github.com/AyasdiOpenSource/bigdf/blob/master/src/main/scala/com/ayasdi/bigdf/SchemaUtils.scala this is what i did :-) an option to use sampling instead of using the whole dataset would be nice. i am happy to extract that code from bigdf and contribute

possible optimizations:
1) let inferField know what we have so far, so it can avoid doing unnecessary work. e.g. if we got a double, no need to try parsing as int or long. if we got a string, no need to parse as double.
2) use mappartitions to do an "imperative-style" loop per partition and then merge the partitions (which can be done on the driver as there will be a single schema per partition). this will reduce memory allocation and GC overhead.

falaki replied:
I like both suggestions. I suggest doing both, although most of the benefit will come from number 2.

falaki commented on src/main/scala/com/databricks/spark/csv/util/InferSchema.scala:
```diff
@@ -0,0 +1,73 @@
+package com.databricks.spark.csv.util
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.catalyst.analysis.HiveTypeCoercion
```
This is not a public API and using it will cause problems if Spark SQL changes. Would you include a simple version of it in your patch? You can copy/paste that code as well. CC @yhuai for comment.

falaki commented on src/main/scala/com/databricks/spark/csv/CsvRelation.scala:
```diff
@@ -88,15 +87,52 @@ case class CsvRelation protected[spark] (
+        parseCSV(csvIter, csvFormat)
+      }
+    }
+
```
extra line.

falaki commented on src/main/scala/com/databricks/spark/csv/CsvRelation.scala:
```diff
@@ -88,15 +87,52 @@ case class CsvRelation protected[spark] (
+
+  // By making this a lazy val we keep the RDD around, amortizing the cost of locating splits.
+  def buildScan = {
+
```
extra line.

falaki commented on src/main/scala/com/databricks/spark/csv/CsvRelation.scala:
```diff
@@ -88,15 +87,52 @@ case class CsvRelation protected[spark] (
+        }
       }
     }
+
```
remove extra line.

falaki commented on src/main/scala/com/databricks/spark/csv/util/InferSchema.scala:
```diff
@@ -0,0 +1,73 @@
+    }
+
+  }
+
```
extra line.

falaki commented on src/main/scala/com/databricks/spark/csv/util/InferSchema.scala:
```diff
@@ -0,0 +1,73 @@
+   *     3. Replace any null types with string type
+   */
+  def apply(
+             tokenRdd: RDD[Array[String]],
```
Is more readable in a single line.

falaki commented:
Thanks @dtpeacock for submitting this. We really need this feature. I am somewhat concerned with using internal Spark API. I also left a few minor style related comments. @mohitjaggi suggestions for optimization are also good to have. I would leave the extra option for sampling for a separate PR.

dtpeacock commented:
Thanks for the comments - I have fixed the style issues and pulled the internal spark method across.

Mapping over each partition and then merging each partition's results is definitely a much better way of doing it - I've replaced the original method with an aggregate which does exactly that. 

As suggested, I've also passed through to inferField the type so far, and skipped unneeded cases, although there might be a better way to implement it.

JoshRosen commented:
I'm experimenting with integrating a code coverage tool into the spark-csv build.  In order to trigger your PR to be re-build with that tool, I'm going to temporarily close it, then will re-open it in order to convince Travis to rebuild the new merge commit.

JoshRosen commented:
Cool, it worked!  The current build status is a little overzealous relative to our current coverage, so I'm going to disable it for now so as not to get false-positive build failures.

JoshRosen commented on src/main/scala/com/databricks/spark/csv/util/InferSchema.scala:
```diff
@@ -0,0 +1,127 @@
+    case (t1, NullType) => Some(t1)
+
+    // Promote numeric types to the highest of the two and all numeric types to unlimited decimal
+    case (t1, t2) if Seq(t1, t2).forall(numericPrecedence.contains) =>
```
Codecov claims that this case and all of the following cases are aren't covered in tests.

JoshRosen commented on src/main/scala/com/databricks/spark/csv/CsvParser.scala:
```diff
@@ -86,6 +87,11 @@ class CsvParser {
     this
   }
 
+  def withInferSchema(inferSchema: Boolean) = {
```
This isn't called anywhere (Codecov caught this).

JoshRosen commented on src/main/scala/com/databricks/spark/csv/CsvRelation.scala:
```diff
@@ -88,7 +87,39 @@ case class CsvRelation protected[spark] (
+          case aiob: ArrayIndexOutOfBoundsException if permissive =>
+            (index until schemaFields.length).foreach(ind => rowArray(ind) = null)
+            Some(Row.fromSeq(rowArray))
+          case NonFatal(e) if !failFast =>
```
This error-handling case is never hit in unit tests.

mohitjaggi commented on src/main/scala/com/databricks/spark/csv/util/InferSchema.scala:
```diff
@@ -0,0 +1,127 @@
+  }
+
+  private[csv] def inferRowType(rowSoFar: Array[DataType], next: Array[String]) = {
+    rowSoFar.zipAll(next, NullType, null).map{case (tpe, field) => inferField(tpe, field)}
```
the inferField enhancement looks ok...however i am not sure if inferRowType() will avoid the GC issue as zipAll and map will create temp objects. rdd.aggregate() allows you to have a "mutable" left param...so you could do something like this:

var i = 0
while(i < rowSoFar.length) {
 rowSoFar(i) = inferField(rowSoFar(i), field) // not sure if findTightestCommonType needs  be called here?
}

JoshRosen commented on src/main/scala/com/databricks/spark/csv/util/InferSchema.scala:
```diff
@@ -0,0 +1,127 @@
+private[csv] object InferSchema {
+
+  /**
+   * Similar to the JSON schema inference. {@link org.apache.spark.sql.json.InferSchema}
```
Will `@link` render correctly in Scaladoc?

dtpeacock commented:
Codecov is pretty nice!

I've removed the unused code, and increased the test coverage.

@mohitjaggi  - does that look better? That should avoid the gc issue now.

mohitjaggi commented:
@dtpeacock optimization itself lgtm. from a code factoring perspective: is there a way to have the "merging" logic in a single place yet have the optimization of avoiding unnecessary exception catching?

falaki commented:
Thanks @dtpeacock merging this.
